File: python_oo/md.py
"""
Object orientated (OO) Python implementation
"""
from math import sqrt


class _3DVector:

    def __init__(self,
                 x: float = 0.0,
                 y: float = 0.0,
                 z: float = 0.0):
        """Vector in 3D space, with x, y, z components"""

        self._x = x
        self._y = y
        self._z = z

    @property
    def x(self) -> float:
        return self._x

    @x.setter
    def x(self, value):
        self._x = float(value)

    @property
    def y(self) -> float:
        return self._y

    @y.setter
    def y(self, value):
        self._y = float(value)

    @property
    def z(self) -> float:
        return self._z

    @z.setter
    def z(self, value):
        self._z = float(value)

    def zero(self) -> None:
        self._x, self._y, self.z = 0.0, 0.0, 0.0

    def __iadd__(self, other: '_3DVector') -> '_3DVector':
        """Inplace addition of another vector to this one"""
        self._x += other.x
        self._y += other.y
        self._z += other.z

        return self

    def __add__(self, other: '_3DVector') -> '_3DVector':
        """Addition of another vector to this one"""
        x0, y0, z0 = self._x, self._y, self._z
        x1, y1, z1 = other.x, other.y, other.z

        return self.__class__(x=x0+x1, y=y0+y1, z=z0+z1)

    def __mul__(self, other: float) -> '_3DVector':
        """Multiply this vector by a scalar"""
        x, y, z = self._x, self._y, self._z

        return self.__class__(x=other*x, y=other*y, z=other*z)

    def __repr__(self):
        return f'{self.__class__.__name__}({self._x}, {self._y}, {self._z})'


class Position(_3DVector):
    """Position vector in 3D space: R"""


class Velocity(_3DVector):
    """Velocity vector in 3D space: dR/dt"""


class Force(_3DVector):
    """Force vector in 3D space: -dE/dx, -dE/dy, -dE/dz"""


class Acceleration(_3DVector):
    """Acceleration vector: dv/dt"""


class _PostiveFloat(float):

    def __new__(cls, value):
        if float(value) < 0:
            raise ValueError(f'{cls.__name__} must be positive. Had: {value}')

        return float.__new__(cls, value)


class Mass(_PostiveFloat):
    """Mass (weight)"""


class Timestep(_PostiveFloat):
    """Time-step (dt)"""


class _PostiveInteger(int):

    def __new__(cls, value):
        if int(value) <= 0:
            raise ValueError(f'{cls.__name__} must be positive. Had: {value}')

        return int.__new__(cls, value)


class NumberOfSimulationSteps(_PostiveInteger):
    """n_steps"""


class Particle:

    def __init__(self):
        """Particle initialised with a default position and velocity"""

        self.mass = Mass(1.0)
        self.position = Position()
        self.velocity = Velocity()

        self._prev_force = Force()
        self._force = Force()

    @property
    def force(self) -> Force:
        """Current force"""
        return self._force

    @force.setter
    def force(self, value: Force) -> None:
        """Set a new value of the force"""
        self._prev_force = self._force
        self._force = value

    @property
    def acceleration(self) -> Acceleration:
        """Acceleration derived using F = ma"""
        a = Acceleration(x=self._force.x / self.mass,
                         y=self._force.y / self.mass,
                         z=self._force.z / self.mass)
        return a

    @property
    def prev_acceleration(self) -> Acceleration:
        """Acceleration derived using F = ma"""
        a = Acceleration(x=self._prev_force.x / self.mass,
                         y=self._prev_force.y / self.mass,
                         z=self._prev_force.z / self.mass)
        return a

    def update_position(self, dt: Timestep) -> None:
        """Update the position according to a velocity verlet update"""
        self.position += self.velocity * dt + self.acceleration * dt**2

    def update_velocity(self, dt: Timestep) -> None:
        """Update the position according to a velocity verlet update"""
        self.velocity += (self.acceleration + self.prev_acceleration) * (dt / 2.)


class PairwisePotential:

    def force(self, particle_i: Particle, particle_j: Particle) -> Force:
        raise NotImplementedError


class LennardJones(PairwisePotential):
    """
    Lennard Jones potential, defined by the energy function:

         E = Σ  4ε ((σ/r_ij)^12 - (σ/r_ij)^6)
            ij

    where i, j enumerate over all particles.
    """

    def __init__(self,
                 sigma:   float,
                 epsilon: float):

        self.sigma = sigma
        self.epsilon = epsilon

        # Required factors for the gradient, evaluated once.
        self._f = [epsilon/2.0, 12 * sigma**12, -6 * sigma**6]

    def force(self,
              particle_i: Particle,
              particle_j: Particle
              ) -> Force:
        """Calculate the force on particle i due to particle j"""

        if id(particle_i) == id(particle_j):  # No self interaction
            return Force(0.0, 0.0, 0.0)

        dx = particle_i.position.x - particle_j.position.x
        dy = particle_i.position.y - particle_j.position.y
        dz = particle_i.position.z - particle_j.position.z
        r = sqrt(dx * dx + dy * dy + dz * dz)

        const = self._f[0] * (self._f[1] * r**(-14) + self._f[2] * r**(-8))

        return Force(x=const * dx, y=const * dy, z=const * dz)


class Particles(list):

    def __init__(self,
                 positions_filename:  str,
                 velocities_filename: str):
        """List of particles constructed from positions and velocities file"""
        super().__init__()

        for _ in range(self._n_particles(positions_filename)):
            self.append(Particle())

        self._set_all('position', positions_filename)
        self._set_all('velocity', velocities_filename)

    def calculate_forces(self,
                         potential: PairwisePotential
                         ) -> None:
        """Set the force on each particle due to an interaction potential"""

        for particle_i in self:
            particle_i.force.zero()

            for particle_j in self:
                particle_i.force += potential.force(particle_i, particle_j)

        return None

    def print_xyz_file(self,
                       filename: str,
                       append:   bool = False
                       ) -> None:
        """Print an xyz file of the particles"""

        with open(filename, 'a' if append else 'w') as xyz_file:

            print(f'{len(self)}\n', file=xyz_file)
            for particle in self:

                pos = particle.position
                print(f'H  {pos.x:.5f}  {pos.y:.5f}  {pos.z:.5f}',
                      file=xyz_file)

        return None

    def print_positions(self,
                        filename: str = 'positions.txt'
                        ) -> None:
        """Print the positions to a .txt file"""

        with open(filename, 'w') as pos_file:
            for particle in self:
                pos = particle.position
                print(f'{pos.x:.5f}  {pos.y:.5f}  {pos.z:.5f}', file=pos_file)

        return None

    def _set_all(self,
                 attr:     str,
                 filename: str
                 ) -> None:
        """
        Set the positions or velocities of all the particles from a file.
        Expecting a file format of:

            x0   y0   z1
            x1   y1   z1
            .    .    .

        where x0 etc. are floating point numbers.

        -----------------------------------------------------------------------
        Arguments:
            attr: Attribute to set. One of {'position', 'velocity'}

            filename:

        Raises:
            (IOError): If the file does not exist

            (ValueError): If the file is malformatted
        """
        for i, line in enumerate(open(filename,  'r')):

            if self._is_blank(line):
                break

            try:
                vec = getattr(self[i], attr)
                vec.x, vec.y, vec.z = line.split()

            except ValueError:
                raise ValueError(f'Could not convert {line} to x, y, z '
                                 f'components')

            except IndexError:
                raise ValueError(f'Could not set position for particle {i} '
                                 f'only had {len(self)} particles!')

        return None

    @staticmethod
    def _is_blank(line: str) -> bool:
        return len(line.split()) == 0

    @staticmethod
    def _n_particles(filename: str) -> int:
        """Number of particles present in a file"""
        return sum(len(line.split()) == 3 for line in open(filename, 'r'))


class Simulation:
    """Simulation of a set of particles"""

    def __init__(self,
                 particles:        Particles,
                 potential:        PairwisePotential,
                 timestep:         Timestep,
                 n_steps:          NumberOfSimulationSteps,
                 print_trajectory: bool = False):
        """Construct a simulation from a set of properties"""

        self.particles = particles

        self._potential = potential
        self._timestep = timestep
        self._n_steps = n_steps
        self._print_trajectory = bool(print_trajectory)

    def run(self) -> None:
        """Run molecular dynamics"""

        if self._print_trajectory:
            self._clear_trajectory_file()

        self.particles.calculate_forces(self._potential)

        for _ in range(self._n_steps):

            if self._print_trajectory:
                self.particles.print_xyz_file(filename='trajectory.xyz',
                                              append=True)

            self._update_positions()
            self.particles.calculate_forces(self._potential)
            self._update_velocities()

        return None

    def _update_positions(self) -> None:
        """Update the position of all the particles in the simulation"""

        for particle in self.particles:
            particle.update_position(dt=self._timestep)

        return None

    def _update_velocities(self) -> None:
        """Update the velocity of all the particles in the simulation"""

        for particle in self.particles:
            particle.update_velocity(dt=self._timestep)

        return None

    @staticmethod
    def _clear_trajectory_file() -> None:
        open('trajectory.xyz', 'w').close()


if __name__ == '__main__':

    cluster = Particles(positions_filename='data/positions.txt',
                        velocities_filename='data/velocities.txt')

    simulation = Simulation(particles=cluster,
                            potential=LennardJones(epsilon=100,
                                                   sigma=1.7),
                            n_steps=NumberOfSimulationSteps(10000),
                            timestep=Timestep(0.01))
    simulation.run()
    simulation.particles.print_positions()

File: cpp_oo/md.cpp
//
// Created by Tom Young on 19/01/2022.
//
#include <string>
#include <utility>
#include <vector>
#include <array>
#include <sstream>
#include <fstream>
#include <iomanip>
#include <cassert>
#include <stdexcept>
#include <cmath>


using namespace std;


vector<string> split(const string &s, char delim) {
    // Split a string by a delimiter into a vector of strings

    stringstream stream(s);
    string item;
    vector<string> elems;
    while (getline(stream, item, delim)) {
        if (!item.empty()){
            elems.push_back(item);
        }
    }
    return elems;
}


template <class T>
class PositiveDouble{

    protected:
        double value = 0.0;

    public:

        PositiveDouble() = default;

        explicit PositiveDouble(double value){
            this->value = value;
            assert(value > 0.0);
        }

        T operator*(double x) const {return T(value*x);}
        T operator/(double x) const {return T(value/x);}

        friend double operator*(double x, T y){return x * y.value;}
        friend double operator/(double x, T y){return x / y.value;}
};


class TimeIncrement: public PositiveDouble<TimeIncrement>{
    // A positive time increment (∆t)

    public:
        TimeIncrement(): PositiveDouble(0.01){};
        explicit TimeIncrement(double value): PositiveDouble(value){};
};


class Mass: public PositiveDouble<Mass>{
    // Mass of a particle

    public:
        Mass() : PositiveDouble(1.0){};
        explicit Mass(double value): PositiveDouble(value){};
};


class PositiveInteger{

    protected:
        int value;

    public:
        explicit PositiveInteger(int value){
            assert(value > 0);
            this->value = value;
        }

        friend bool operator<(int i, PositiveInteger j) {
            return  i < j.value;
        }
};


class NumberOfTimeSteps: public PositiveInteger{
    public:
        NumberOfTimeSteps() : PositiveInteger(1){};
        explicit NumberOfTimeSteps(int value): PositiveInteger(value){};
};


template <class T>
class Vector3D: public array<double, 3>{
    // A vector in 3D space with x, y, z components

    public:

        double x() const {return at(0);}
        double y() const {return at(1);}
        double z() const {return at(2);}

        void zero(){
            this->at(0) = 0.0;
            this->at(1) = 0.0;
            this->at(2) = 0.0;
        }

        T clone(){return {x(), y(), z()};}

        friend ostream &operator<<(std::ostream &os, Vector3D const &vec) {
            // Overloaded << operator for printing a vector
            return os << '(' << vec.x() << ", " << vec.y() << ", " << vec.z() << ')';
        }

        T operator+=(const T& other){
            // Add another position to this one

            // TODO: Work out how to do this without a copy
            this->at(0) += other.x();
            this->at(1) += other.y();
            this->at(2) += other.z();

            return {this->at(0), this->at(1), this->at(2)};
        }

        T operator+(const T& other){
            // Add another position to this one
            return {x()+other.x(), y()+other.y(), z()+other.z()};
        }
};


class Position: public Vector3D<Position>{};


class Velocity: public Vector3D<Velocity>{

    public:
        Position operator*(TimeIncrement dt){
            // r ≈ dr/dt ∆t = v ∆t
            return {x() * dt, y() * dt, z() * dt};
        }
};


class Acceleration: public Vector3D<Acceleration>{

    public:
        Velocity operator*(TimeIncrement dt) const{
            // dr/dt ≈ d^2r/dt^2 ∆t = a ∆t
            return {x() * dt, y() * dt, z() * dt};
        }
};


class Force: public Vector3D<Force>{

    public:
        Acceleration operator/(Mass& m) const {
            return {x() / m, y() / m, z() / m};
        }
};


class Particle{
    // Particle with a defined positions, velocity and force

    public:
        Mass mass = Mass(1.0);
        Position position = {0.0, 0.0, 0.0};
        Velocity velocity = {0.0, 0.0, 0.0};
        Force force = {0.0, 0.0, 0.0};
        Force prev_force = {0.0, 0.0, 0.0};

        Particle(double x, double y, double z){
            // Initialise a particle in a defined position

            this->position = {x, y, z};
        }

        void update_position(TimeIncrement dt){
            // Update the position based on a velocity verlet
            position += (velocity * dt) + (a() * dt * dt);
        }

        void update_velocity(TimeIncrement dt){
            velocity += (a() + a_prev()) * (dt / 2.);
        }

        friend bool operator==(const Particle& lhs, const Particle& rhs){
            // Equality pf particles is defined by equality of references
            return &lhs == &rhs;
        }

    protected:

        Acceleration a(){
            // Acceleration due to the current force on the particle (F = ma)
            return force / mass;
        }

        Acceleration a_prev(){
            // Acceleration due to the previous force on the particle
            return prev_force / mass;
        }

};


class LJPotential{

    public:
        LJPotential() = default;

        LJPotential(double epsilon, double sigma){
            // Initialise a Lennard Jones potential from ε and σ

            f[0] = epsilon / 2.0;
            f[1] = 12 * pow(sigma, 12);
            f[2] = -6 * pow(sigma, 6);
        }

        void add_force(Particle &particle_i,
                       Particle &particle_j){
            // Add the force on particle i due to particle j

            if (particle_i == particle_j) return;  // No self interaction

            double dx = particle_i.position.x() - particle_j.position.x();
            double dy = particle_i.position.y() - particle_j.position.y();
            double dz = particle_i.position.z() - particle_j.position.z();
            double r = sqrt(dx * dx + dy * dy + dz * dz);

            double c = f[0] * (f[1] * pow(r, -14) + f[2] * pow(r, -8));

            particle_i.force[0] += c * dx;
            particle_i.force[1] += c * dy;
            particle_i.force[2] += c * dz;
        }

    protected:
        // Coefficients used to calculate the gradient
        array<double, 3> f = {0.0, 0.0, 0.0};

};


class Particles : public vector<Particle>{
    // List of particles

    public:
        Particles() = default;

        explicit Particles(const string& positions_filename){
            // Initialise particles from a file of x,y,z positions

            add_particles(positions_filename);
        }

        void set_velocities(const string& filename){
            // Set velocities from a file of x,y,z velocities
            int i = 0;

            for (auto &vec : matrix_from(filename)){

                if (i > this->size()){
                    throw runtime_error("Cannot set velocity for particle");
                }
                this->data()[i].velocity = {vec[0], vec[1], vec[2]};

                i++;
            }
        }

        void calculate_forces(LJPotential &potential){
            // Calculate the forces on all particles

            for (auto &particle_i : *this){

                particle_i.prev_force = particle_i.force.clone();
                particle_i.force.zero();

                for (auto &particle_j: *this){
                    potential.add_force(particle_i, particle_j);
                }
            }
        }

        void print_positions(const string& filename){
            // Print the positions to a file

            ofstream file;
            file.open(filename);

            for (auto &particle : *this){
                file << setprecision(8)
                     << particle.position.x() << " "
                     << particle.position.y() << " "
                     << particle.position.z() << "\n";
            }

            file.close();
        }

        void update_positions(const TimeIncrement& dt){
            // Update the positions of all the particles

            for (auto &particle : *this){
                particle.update_position(dt);
            }
        }

        void update_velocities(const TimeIncrement& dt){
            // Update the velocities of all the particles

            for (auto &particle : *this){
                particle.update_velocity(dt);
            }
        }

    protected:
        void add_particles(const string& filename){
            // Add particles at defined positions from a file

            for (auto &vec : matrix_from(filename)){
                emplace_back(vec[0], vec[1], vec[2]);
            }
        }

        static vector<array<double, 3>> matrix_from(const string& filename){
            // Extract a Nx3 matrix of strings from a file

            string line;
            ifstream _file(filename);
            vector<array<double, 3>> matrix;

            while (getline(_file, line, '\n')) {

                if (line.empty()) continue;

                vector<string> xyz_items = split(line, ' ');

                if (xyz_items.size() != 3){
                    throw runtime_error("xyz line was not of the correct "
                                        "format. Must be (x y z), had: "+line);
                }
                matrix.push_back({stod(xyz_items[0]),
                                  stod(xyz_items[1]),
                                  stod(xyz_items[2])});
            }
            _file.close();

            return matrix;
        }

};


class Simulation{

    protected:
        LJPotential potential;
        NumberOfTimeSteps n_steps;
        TimeIncrement dt;

    public:
        Particles particles;

        Simulation(Particles _particles,
                   LJPotential _potential,
                   NumberOfTimeSteps _n_steps,
                   TimeIncrement _dt){
            // Initialise a simulation

            this->particles = move(_particles);
            this->potential = _potential;
            this->n_steps = _n_steps;
            this->dt = _dt;
        }

        void run(){
            // Run the simulation using a velocity verlet update

            particles.calculate_forces(potential);

            for (int i = 0; i < n_steps; i++){
                particles.update_positions(dt);
                particles.calculate_forces(potential);
                particles.update_velocities(dt);
            }
        }

};


int main(){

    auto cluster = Particles("positions.txt");
    cluster.set_velocities("velocities.txt");

    auto simulation = Simulation(cluster,
                                 LJPotential(100.0, 1.7),
                                 NumberOfTimeSteps(10000),
                                 TimeIncrement(0.01));
    simulation.run();

    simulation.particles.print_positions("final_positions.txt");

    return 0;
}
